File: macos/QMK Toolbox/KeyTester/KeyView.m
#import "KeyView.h"

@interface KeyView ()
@property (weak) IBOutlet NSView *contentView;

@property (weak) IBOutlet NSBox *boxView;

@property (weak) IBOutlet NSTextField *legendView;
@end

@implementation KeyView
- (instancetype)initWithFrame:(NSRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self customInit];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        [self customInit];
    }
    return self;
}

- (void)setPressed:(BOOL)pressed {
    _pressed = pressed;
    [self setNeedsDisplay:YES];
}

- (void)setTested:(BOOL)tested {
    _tested = tested;
    [self setNeedsDisplay:YES];
}

- (NSString *)legend {
    return self.legendView.stringValue;
}

- (void)setLegend:(NSString *)legend {
    self.legendView.stringValue = legend;
    [self setNeedsDisplay:YES];
}

- (void)drawRect:(NSRect)dirtyRect {
    [super drawRect:dirtyRect];

    self.boxView.fillColor = self.pressed ? [NSColor systemYellowColor] : self.tested ? [NSColor systemGreenColor] : nil;
    self.legendView.textColor = (self.pressed || self.tested) ? [NSColor blackColor] : [NSColor secondaryLabelColor];
}

- (void)customInit {
    [[NSBundle bundleForClass:self.class] loadNibNamed:@"KeyView" owner:self topLevelObjects:nil];

    NSMutableArray<NSLayoutConstraint *> *newConstraints = [[NSMutableArray alloc] init];

    for (NSLayoutConstraint *oldConstraint in self.contentView.constraints) {
        NSView *firstItem = [oldConstraint.firstItem isEqualTo:self.contentView] ? self : oldConstraint.firstItem;
        NSView *secondItem = [oldConstraint.secondItem isEqualTo:self.contentView] ? self : oldConstraint.secondItem;

        [newConstraints addObject:[NSLayoutConstraint constraintWithItem:firstItem attribute:oldConstraint.firstAttribute relatedBy:oldConstraint.relation toItem:secondItem attribute:oldConstraint.secondAttribute multiplier:oldConstraint.multiplier constant:oldConstraint.constant]];

        for (NSView *newView in self.contentView.subviews) {
            [self addSubview:newView];
        }

        [self addConstraints:newConstraints];
    }
}
@end

File: macos/QMK Toolbox/KeyTester/KeyView.swift
import AppKit

@IBDesignable
class KeyView: NSView {
    @IBOutlet var contentView: NSView!
    @IBOutlet var boxView: NSBox!
    @IBOutlet var legendView: NSTextField!

    @IBInspectable
    var pressed: Bool = false {
        didSet {
            needsDisplay = true
        }
    }

    @IBInspectable
    var tested: Bool = false {
        didSet {
            needsDisplay = true
        }
    }

    @IBInspectable
    var legend: String = "" {
        didSet {
            legendView.stringValue = legend
            needsDisplay = true
        }
    }

    override init(frame frameRect: NSRect) {
        super.init(frame: frameRect)
        customInit()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        customInit()
    }

    override func draw(_ dirtyRect: NSRect) {
        super.draw(dirtyRect)
        boxView.fillColor = pressed ? NSColor.systemYellow : tested ? NSColor.systemGreen : NSColor.clear
        legendView.textColor = (pressed || tested) ? NSColor.black : NSColor.secondaryLabelColor
    }

    func customInit() {
        let bundle = Bundle(for: type(of: self))
        let nib = NSNib(nibNamed: "KeyView", bundle: bundle)!
        nib.instantiate(withOwner: self, topLevelObjects: nil)
        addSubview(contentView)

        var newConstraints: [NSLayoutConstraint] = []
        contentView.constraints.forEach { oldConstraint in
            let firstItem: NSView = oldConstraint.firstItem!.isEqual(to: contentView) ? self : oldConstraint.firstItem as! NSView
            let secondItem: NSView = oldConstraint.secondItem!.isEqual(to: contentView) ? self : oldConstraint.secondItem as! NSView

            newConstraints.append(NSLayoutConstraint(item: firstItem, attribute: oldConstraint.firstAttribute, relatedBy: oldConstraint.relation, toItem: secondItem, attribute: oldConstraint.secondAttribute, multiplier: oldConstraint.multiplier, constant: oldConstraint.constant))

            contentView.subviews.forEach { newView in
                addSubview(newView)
            }

            addConstraints(newConstraints)
        }
    }
}
